FreeRTOS/FreeRTOS-Kernel
# RoundRobin in task with same priority when configUSE_TIME_SLICING = 0

Fix: RoundRobin in task with same priority when configUSE_TIME_SLICING = 0
---------------------------------------------------------------------------------------------------------

Description:
---------------------------------------------------------------------------------------------------------
As we know, there are several possible scheduling policies. Some of them FIFO, Round Robin, Fixed-priority preemptive and others. By default, freeRTOS uses the Prioritized Preemptive Scheduling with Time Slicing policy. However, it is possible to modify the scheduler policy with few changes in code. Theoretically, it is only necessary to change two defines, namely configUSE_PREEMPTION and configUSE_TIME_SLICING.

However, if by chance we wanted to use the Fixed-priority preemptive policy, we would have to allow a change of context only when a higher priority task appears, that is, tasks of equal priorities will not share processing time. This was to be configured with configUSE_TIME_SLICING. Therefore, we need our scheduler to be preemptive and not always go to the next task if they have the same priority.

Therefore, we can check the information provided by freeRTOS. Following is an excerpt taken from the website[1].

"configUSE_TIME_SLICING:
By default (if configUSE_TIME_SLICING is not defined, or if configUSE_TIME_SLICING is defined as 1) FreeRTOS uses prioritized preemptive scheduling with time slicing. That means the RTOS scheduler will always run the highest priority task that is in the Ready state, and will switch between tasks of equal priority on every RTOS tick interrupt. If configUSE_TIME_SLICING is set to 0 then the RTOS scheduler will still run the highest priority task that is in the Ready state, but will not switch between tasks of equal priority just because a tick interrupt has occurred. "

However, if you set configUSE_TIME_SLICING = 0, it continues to switch context on tasks of equal priority. To solve this, it was necessary to add a check to see if the time_slicing is set or not, to decide which task will be performed, between it or next.

Test Steps:
---------------------------------------------------------------------------------------------------------
1° Create two tasks with same priority and print something many times with a delay between them.
2° Check if occur any context switch, seeing the debugs prints.
3° Check if the scheduler work in FIFO policy.

Example of task utilized:
---------------------------------------------------------------------------------------------------------
```c
inline void delay(){
        unsigned int i;
        for(i = 0; i < 800000000; i++) {
                asm("nop");
        }
}

void taskLed(void *pvParameters){
        int led = 0, i = 0;
        char *name = (char *)pvParameters;
        for (i = 0;i < 10; i++) {
                if(led){
                        printf("Led %s ON\n" , name);
                }else{
                        printf("Led %s OFF\n", name);
                }
                led ^= 1, delay();
        }
        vTaskDelete(NULL);
}
```

Related Issue:
---------------------------------------------------------------------------------------------------------

https://www.freertos.org/FreeRTOS_Support_Forum_Archive/December_2013/freertos_configUSE_TIME_SLICING_278999daj.html

http://www.openrtos.net/FreeRTOS_Support_Forum_Archive/December_2015/freertos_Tasks_w_same_Priority_Interrupt_even_with_time_slice_off_9336ef42j.html

https://www.freertos.org/FreeRTOS_Support_Forum_Archive/February_2018/freertos_Preemption_by_an_equal_priority_task_c8cb93d0j.html

https://www.freertos.org/FreeRTOS_Support_Forum_Archive/March_2019/freertos_In_the_case_of_configUSE_TIME_SLICING_0_aaaec33df6j.html

References:
---------------------------------------------------------------------------------------------------------
[1]https://www.freertos.org/a00110.html

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
